diff --git a/stablehlo/transforms/CMakeLists.txt b/stablehlo/transforms/CMakeLists.txt
index 4787369d..88d08012 100644
--- a/stablehlo/transforms/CMakeLists.txt
+++ b/stablehlo/transforms/CMakeLists.txt
@@ -39,6 +39,22 @@ set(LLVM_TARGET_DEFINITIONS VhloToVersionPatterns.td)
 mlir_tablegen(VhloToVersionPatterns.h.inc --gen-rewriters)
 add_public_tablegen_target(VhloToVersionPatterns)
 
+add_mlir_dialect_library(StablehloPassUtils
+  PARTIAL_SOURCES_INTENDED
+  PassUtils.cpp
+
+  DEPENDS
+  PassesIncGen
+
+  LINK_LIBS PUBLIC
+  ChloOps
+  StablehloBase
+  StablehloOps
+  MLIRSupport
+  MLIRComplexDialect
+  MLIRIR
+)
+
 
 add_mlir_dialect_library(StablehloPasses
   PARTIAL_SOURCES_INTENDED
@@ -59,7 +75,6 @@ add_mlir_dialect_library(StablehloPasses
   StablehloRefineShapes.cpp
   VhloLegalizeToStablehlo.cpp
   VhloToVersion.cpp
-  PassUtils.cpp
 
   DEPENDS
   ChloDecompositionPatternsIncGen
@@ -68,6 +83,7 @@ add_mlir_dialect_library(StablehloPasses
   StablehloComplexMathExpanderPatternsIncGen
   StablehloLegalizeDeprecatedOpsPatternsIncGen
   VhloToVersionPatterns
+  StablehloPassUtils
 
   LINK_LIBS PUBLIC
   ChloOps
@@ -92,4 +108,5 @@ add_mlir_dialect_library(StablehloPasses
   StablehloOptimizationPasses
   StablehloTypeInference
   VhloOps
+  StablehloPassUtils
 )
diff --git a/stablehlo/transforms/optimization/CMakeLists.txt b/stablehlo/transforms/optimization/CMakeLists.txt
index d43d77be..65d9af94 100644
--- a/stablehlo/transforms/optimization/CMakeLists.txt
+++ b/stablehlo/transforms/optimization/CMakeLists.txt
@@ -29,6 +29,7 @@ add_mlir_dialect_library(StablehloOptimizationPasses
   DEPENDS
   OptimizationPassesIncGen
   StablehloAggressiveSimplificationPatternsIncGen
+  StablehloPassUtils
 
   LINK_LIBS PUBLIC
   ChloOps
@@ -42,4 +43,5 @@ add_mlir_dialect_library(StablehloOptimizationPasses
   StablehloBase
   StablehloOps
   StablehloTypeInference
+  StablehloPassUtils
 )
